tiledb/sm/filesystem/ssl_config.cc

```diff
@@ -0,0 +1,110 @@
+/**
+ * @file ssl_config.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file includes definitions of the SSLConfig class.
+ */
+
+#include "tiledb/sm/filesystem/ssl_config.h"
+#include "tiledb/platform/cert_file.h"
+
+namespace tiledb::sm {
+
+SSLConfig::SSLConfig()
+    : ca_file_("")
+    , ca_path_("")
+    , verify_(true) {
+}
+
+SSLConfig::SSLConfig(const Config& cfg)
+    : ca_file_("")
+    , ca_path_("")
+    , verify_(true) {
+  // Look up our ca_file and ca_path configuration options
+  auto ca_file = cfg.get<std::string>("ssl.ca_file");
+  if (ca_file.has_value()) {
+    ca_file_ = ca_file.value();
+  }
+
+  auto ca_path = cfg.get<std::string>("ssl.ca_path");
+  if (ca_path.has_value()) {
+    ca_path_ = ca_path.value();
+  }
+
+  if constexpr (tiledb::platform::PlatformCertFile::enabled) {
+    // If neither ca_file or ca_path are set, we look for a system default
+    // CA file on Linux platforms.
+    if (ca_file_.empty() && ca_path_.empty()) {
+      ca_file_ = tiledb::platform::PlatformCertFile::get();
+    }
+  }
+
+  auto verify = cfg.get<bool>("ssl.verify");
+  if (verify.has_value()) {
+    verify_ = verify.value();
+  }
+}
+
+S3SSLConfig::S3SSLConfig(const Config& cfg)
+    : SSLConfig(cfg) {
+  // Support the old s3 configuration values if they are
+  // configured by the user.
+
+  // Only set ca_file_ if vfs.s3.ca_file is a non-empty string
+  auto ca_file = cfg.get<std::string>("vfs.s3.ca_file");
```

Comment:
Do we have some sort of deprecation mechanism for these config options? Thinking from the user's perspective, when I'd search for options documentation/forums/slack/etc, I'd definitely get a few hits with the old options and I'd get pretty confused on which one of the ssl options would help me. I'm wondering whether we should signal somehow that those options are not to be used anymore (warning/docs comments).

Reply:
I see your point, and I have absolutely no idea on the docs side of things. I'd agree that at least adding notes to the s3 specific config variable docs would be useful though. As for how long until we drop the s3 variables, I'd assume that's a 3.x breaking change type of event.